### PR TITLE
ENH: DPCTL native support

### PR DESCRIPTION
**Pull request by KulikovNikita**

### CODE REVIEW

File: onedal/datatypes/_data_conversion.py [hunk 1]
```diff
@@ -25,6 +32,10 @@ def from_table(*args):
 
 
 def convert_one_to_table(arg):
+    if dpctl_available:
+        ndarr = dpt.usm_ndarray
+        if isinstance(arg, ndarr):
```
Comment (samir-nasibli) [inline onedal/datatypes/_data_conversion.py:37]: What if dpctl is available and the passed arg is not usm_ndarray?

File: onedal/datatypes/data_conversion_dpctl.cpp [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    // need "version", "data", "shape", "typestr", "syclobj"
+    py::tuple shape = py::make_tuple(static_cast<npy_intp>(homogen_input.get_row_count()), static_cast<npy_intp>(homogen_input.get_column_count()));
+    py::list data_entry(2);
+
+    auto bytes_array = dal::detail::get_original_data(homogen_input);
+    auto queue = bytes_array.get_queue().value();
+
+    // TODO:
+    // size_t
+    data_entry[0] = reinterpret_cast<size_t>(bytes_array.get_data());
```
Comment (KulikovNikita) [inline onedal/datatypes/data_conversion_dpctl.cpp:146]:
```suggestion
    data_entry[0] = reinterpret_cast<std::size_t>(bytes_array.get_data());
```

File: onedal/datatypes/data_conversion_dpctl.cpp [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    if (kind != dal::homogen_table::kind())
+        report_problem_to_dptensor(": only homogen tables are supported");
+
+    const auto& homogen_input = reinterpret_cast<const dal::homogen_table&>(input);
+
+    // need "version", "data", "shape", "typestr", "syclobj"
+    py::tuple shape = py::make_tuple(static_cast<npy_intp>(homogen_input.get_row_count()), static_cast<npy_intp>(homogen_input.get_column_count()));
+    py::list data_entry(2);
+
+    auto bytes_array = dal::detail::get_original_data(homogen_input);
```
Comment (KulikovNikita) [inline onedal/datatypes/data_conversion_dpctl.cpp:141]: I was not aware of this functionality. Thx

File: onedal/datatypes/table.cpp [hunk 4]
```diff
@@ -59,6 +68,20 @@ ONEDAL_PY_INIT_MODULE(table) {
...
+#ifdef ONEDAL_DPCTL_INTEGRATION
+    m.def("dpctl_to_table", [](py::object obj) {
+        return convert_from_dptensor(obj);
+    });
+
+    // We using __sycl_usm_array_interface__ attribute for constructing dpctl
+    // tensor on python level.
+    py::class_<sua_iface_from_table>(m, "get_sua_iface_from_table")
+    .def(py::init([](dal::table& t) { return sua_iface_from_table(construct_sua_iface(t));}))
+    .def_readwrite("__sycl_usm_array_interface__", &sua_iface_from_table::__sycl_usm_array_interface__, py::return_value_policy::take_ownership);
```
[the remark this answers is not in the document]
Comment (samir-nasibli) [reply onedal/datatypes/table.cpp]: it is already done :)

File: onedal/datatypes/_data_conversion.py [hunk 5]
```diff
@@ -15,9 +15,17 @@
...
+from onedal import _is_dpc_backend
 from onedal import _backend
 from daal4py.sklearn._utils import make2d
 
+try:
+    import dpctl
+    import dpctl.tensor as dpt
+    dpctl_available = dpctl.__version__ >= '0.14'
+except ImportError:
+    dpctl_available = False
```
Comment (Alexsandruss) [inline onedal/datatypes/_data_conversion.py:27]: This code is duplicated multiple times across repo. Move it to dedicated file

File: onedal/datatypes/_data_conversion.py [hunk 6]
```diff
@@ -28,8 +36,15 @@ def _apply_and_pass(func, *args):
...
 
+# TODO:
+# refactoring.
+
 
 def convert_one_to_table(arg):
+    if dpctl_available:
+        ndarr = dpt.usm_ndarray
+        if isinstance(arg, ndarr):
+            return _backend.dpctl_to_table(arg)
```
Comment (Alexsandruss) [inline onedal/datatypes/_data_conversion.py:47]:
```suggestion
        if isinstance(arg, dpt.usm_ndarray):
            return _backend.dpctl_to_table(arg)
```

File: onedal/datatypes/data_conversion_dpctl.cpp [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+void report_problem_from_dptensor(const char* clarification) {
+    constexpr const char* const base_message = "Unable to convert from dptensor";
+
+    std::string message{ base_message };
+    message += std::string{ clarification };
+    throw std::invalid_argument{ message };
+}
+
+std::int64_t get_and_check_dptensor_ndim(const dpctl::tensor::usm_ndarray& tensor) {
+    constexpr const char* const err_message = ": only 1D & 2D tensors are allowed";
```
Comment (Alexsandruss) [inline onedal/datatypes/data_conversion_dpctl.cpp:44]: `: ` should be moved to `report_problem_from_dptensor`

File: onedal/datatypes/data_conversion_dpctl.cpp [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+    return res;
+}
+
+void report_problem_to_dptensor(const char* clarification) {
+    constexpr const char* const base_message = "Unable to convert to dptensor";
+
+    std::string message{ base_message };
+    message += std::string{ clarification };
+    throw std::runtime_error{ message };
+}
```
Comment (Alexsandruss) [inline onedal/datatypes/data_conversion_dpctl.cpp:133]: Should be unified with `report_problem_from_dptensor`: for example, add `bool is_convesion_from_dptensor` argument and rename to `report_problem_with_dptensor`

File: onedal/datatypes/data_conversion_dpctl.cpp (hunk 7 above)
Comment (KulikovNikita) [reply onedal/datatypes/data_conversion_dpctl.cpp:44]: @Alexsandruss - I suppose functions like `report_error...` an be called without any conclusive issue

File: onedal/datatypes/_data_conversion.py (hunk 5 above)
Comment (samir-nasibli) [reply onedal/datatypes/_data_conversion.py:29]: Good point. Let me implement it as a feature in other PR.

File: onedal/datatypes/tests/test_data.py [hunk 9]
```diff
@@ -133,3 +142,55 @@ def _test_input_format_f_contiguous_pandas(queue, dtype):
...
+    x_dpt = dpt.asarray(x_numpy, usm_type="device", sycl_queue=queue)
+    # assert not x_dpt.flags.fnc
+    assert isinstance(x_dpt, dpt.usm_ndarray)
+
+    x_table = _backend.dpctl_to_table(x_dpt)
+    assert hasattr(x_table, '__sycl_usm_array_interface__')
+    x_dpt_from_table = dpt.asarray(x_table)
+
+    assert x_dpt.__sycl_usm_array_interface__[
+        'data'][0] == x_dpt_from_table.__sycl_usm_array_interface__['data'][0]
```
Comment (KulikovNikita) [inline onedal/datatypes/tests/test_data.py:191]: Looks like an `autopep8` artifact.

**Review by samir-nasibli — APPROVED** [review]

File: onedal/datatypes/_data_conversion.py (hunk 5 above)
Comment (Alexsandruss) [reply onedal/datatypes/_data_conversion.py:29]: There is no problem to implement it here

**Comment by Alexsandruss** [comment]
@samir-nasibli, @KulikovNikita , my comments were not addressed, please, create PR to fix them.

File: onedal/datatypes/_data_conversion.py (hunk 5 above)
Comment (samir-nasibli) [reply onedal/datatypes/_data_conversion.py:29]: These kind of refactoring is not a purpose of this PR. It should be logically in other  PR. As you mentioned it's duplicated several times, we need some common primitives for that.

**Comment by samir-nasibli** [comment]
> @samir-nasibli, @KulikovNikita , my comments were not addressed, please, create PR to fix them.

Thank you for the comments! #1200

File: onedal/datatypes/_data_conversion.py [hunk 10]
```diff
@@ -29,8 +38,14 @@ def _apply_and_pass(func, *args):
 def from_table(*args):
     return _apply_and_pass(_backend.from_table, *args)
 
+# TODO:
+# refactoring.
```
[the remark this answers is not in the document]
Comment (samir-nasibli) [reply onedal/datatypes/_data_conversion.py:42]: Already done, just forget remove. Thx.

File: onedal/datatypes/tests/test_data.py [hunk 11]
```diff
@@ -133,3 +142,55 @@ def _test_input_format_f_contiguous_pandas(queue, dtype):
...
+    assert x_dpt.shape == x_dpt_from_table.shape
+    assert x_dpt.strides == x_dpt_from_table.strides
+    assert x_dpt.dtype == x_dpt_from_table.dtype
+    assert x_dpt.flags.c_contiguous
+    assert x_dpt_from_table.flags.c_contiguous
+
+
+@pytest.mark.skipif(not dpctl_available,
+                    reason="requires dpctl=0.14")
```
[the remark this answers is not in the document]
Comment (samir-nasibli) [reply onedal/datatypes/tests/test_data.py:174]: #1200

**Pull request merged**